### PR TITLE
ARROW-10545: [C++] Fix crash on invalid Parquet file (OSS-Fuzz)

### PR DESCRIPTION
**Pull request by pitrou**
Also removed a memory allocation (probably not performance-critical).

### CODE REVIEW

File: cpp/src/parquet/level_conversion.cc [hunk 1]
```diff
@@ -35,16 +36,17 @@ namespace internal {
...
                             int64_t num_def_levels, LevelInfo level_info,
                             ValidityBitmapInputOutput* output, OffsetType* offsets) {
   OffsetType* orig_pos = offsets;
-  std::unique_ptr<::arrow::internal::FirstTimeBitmapWriter> valid_bits_writer;
+  optional<::arrow::internal::FirstTimeBitmapWriter> valid_bits_writer;
   if (output->valid_bits) {
-    valid_bits_writer.reset(new ::arrow::internal::FirstTimeBitmapWriter(
-        output->valid_bits, output->valid_bits_offset, num_def_levels));
+    valid_bits_writer.emplace(output->valid_bits, output->valid_bits_offset,
+                              output->values_read_upper_bound);
```
Comment (pitrou) [inline cpp/src/parquet/level_conversion.cc:49]: The bitmap writer length was wrong here. @emkornfield

**Comment by pitrou** [comment]
Note the ASAN CI tests will probably fail until #8617 is merged.

File: cpp/src/parquet/level_conversion.cc (hunk 1 above)
Comment (emkornfield) [reply cpp/src/parquet/level_conversion.cc:49]: :( nice catch.

**Comment by emkornfield** [comment]
LGTM, feel free to merge, or I'll do it when I'm on a computer that i can do it from.

**Comment by pitrou** [comment]
As I said above, this will break CI until #8617 is merged.

**Pull request closed without merging**